### DISCUSSION TITLE
inverter not found, or card not available

### DISCUSSION THREAD

**Discussion by lamixer** [nathanmarlor/foxess_modbus#425] [Support] [answered]
Hello!

Hello -- thanks for all the work on foxess-modbus support -- really amazing to have this! Regarding the set charge period lovelace card:

First I installed the latest stable version, 1.01, which seemed to work fine but when I tried to use the card it said it could not find my inverter, friendly name or ID of ''. Magic menu did not find it either. So I read a few closed issues and though maybe I had to use the beta, so I uninstalled and reinstalled by sliding over to beta and choosing version at the bottom 'main' (my other choices were 1.0 and 1.01).

With the 'main' version installed, if I try to add the card, the card is not present to add.

I have a KH10.5 with modbus integration via rs485. Thanks!

**Comment by canton7** [comment]
Ack, my bad: the README should say that it works with the H1 and AC1, not the KH.

Unfortunately the KH doesn't seem to support setting charge periods over modbus.

Sorry!

**Comment by lamixer** [comment]
Weird! If you need me to poke at it, after giving me a little tutorial, I can investigate mine.

**Reply by canton7** [reply]
We actually have the spec for the KH (https://github.com/nathanmarlor/foxess_modbus/files/11552447/KH.Series.Modbus.pdf). On the H1, the charge period registes start at 41001, but that range in the KH doc is just "TBD".

**Comment by lamixer** [comment]
But maybe that means they haven't gotten around to writing it in the spec, but it actually would work like the other models. Does the card itself try to poke those registers and determine they don't work by the response (or lack thereof)?

**Reply by canton7** [reply]
You're welcome to try and read them with modpoll, see if you can get anything sensible out

**Comment by calum-mcfarlane** [comment]
> But maybe that means they haven't gotten around to writing it in the spec, but it actually would work like the other models. Does the card itself try to poke those registers and determine they don't work by the response (or lack thereof)?

The card can only work with the parameters that the integration presents to it. As the integration doesn't 'know' how to talk to teh KH to set charging times, the card doesn't work. The card is just an interface for the service that allows sending values to the appropriate registers on the inverter.

Now - if you already have the modbus interface set up via RS485, you can easily download modpoll from here https://www.modbusdriver.com/modpoll.html and use it at the command line to see what values exist at those registers on the KH (if any). If you can find values, see if they change when you change the other parameters via the front panel or the app or whatever. If they do - maybe you're onto something.

**Comment by lamixer** [comment]
I'll try modpoll. Since I'm new, and don't really want to destroy my inverter, could you please suggest a command to start with? I'll be running the linux version and talking to the inverter via the modbus-ethernet (Waveshare) box.

**Comment by calum-mcfarlane** [comment]
Have a look here https://github.com/nathanmarlor/foxess_modbus/wiki/USR-TCP232-410s#troubleshooting for a starting point. 

The full list of flags is here https://www.modbusdriver.com/modpoll.html#:~:text=modpoll%20is%20a%20command%20line,Linux%20PCs

(You can't hurt anything as long as you're not trying to write any values!)

**Comment by lamixer** [comment]
Thanks Calum! That is the example I was looking for. First attempt, not so good though!

```
./modpoll -m tcp -p 502 -a 247 -t 3 -0 -1 -c 16 -r 11000 ws-foxess-485.wifi.lan
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 16
Communication.........: ws-foxess-485.wifi.lan, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

Segmentation fault
```

**Comment by lamixer** [comment]
Hmm. I later downloaded it again on my raspberry pi to see if I got any different results running the aarch64 version, but same output.

I guess I should do more reading, but does the 'start reference' mean what register it is reading? KH series looks like the first register used is 30000.

```
root@cs-rpi:~/modpoll/aarch64-linux-gnu# ./modpoll -m tcp -p 502 -a 247 -t 3 -0 -1 -c 16 -r 31016 ws-foxess-485.wifi.lan
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 31016 (PDU), count = 16
Communication.........: ws-foxess-485.wifi.lan, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

Segmentation fault
```

**Comment by lamixer** [comment]
Ah! I'm an idiot. Can't use the hostname have to use IP address. Progress being made!

**Comment by lamixer** [comment]
KH register outputs:

```
root@cs-rpi:~/modpoll/aarch64-linux-gnu# ./modpoll -m tcp -p 502 -a 247 -t 3 -0 -1 -c 16 -r 41000 192.168.1.18
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 41000 (PDU), count = 16
Communication.........: 192.168.1.18, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
[41000]: 0
[41001]: 1
[41002]: 279
[41003]: 546
[41004]: 0
[41005]: 768
[41006]: 1309
[41007]: 260
[41008]: 400
[41009]: 10
[41010]: 100
[41011]: 33
[41012]: 9000
[41013]: 1
[41014]: 1
[41015]: 3
```

I can confirm that 41007 - 41012 represent correct settings.

Current settings for force charge (not ones I actually use) are:

Time Period One
Enable Charge ON
01:23 - 02:34
Time Period Two
Enable Charge OFF
03:00 - 05:29

**Comment by lamixer** [comment]
If I swap my force charge settings to:

Time Period One
Enable Charge ON
12:34 - 13:34
Time Period Two
Enable Charge ON
14:44 - 14:55

I get

```
-- Polling slave...
[41000]: 0
[41001]: 1
[41002]: 3106
[41003]: 3362
[41004]: 1
[41005]: 3628
[41006]: 3639
[41007]: 260
[41008]: 400
[41009]: 10
[41010]: 100
[41011]: 33
[41012]: 9000
[41013]: 1
[41014]: 1
[41015]: 3
```

**Comment by lamixer** [comment]
Looks like it works fine on KH 10.5! Please update the card and I'll try it.

I'm sure there's a logical way to get from 12:34 as a time to 3106 as a value and similar for the others (which you already know). Clearly 41001 and 41004 are whether or not the 'enable charge from grid' setting.

Hey, I think I was helpful for once! :-)

**Reply by canton7** [reply]
Awesome, nice work!

Yep 3106 maps to 12:34. It's `(hours * 256) + minutes`.

I'll get a beta out when I get the chance. It's the integration which will update - the card won't need to change.

**Reply by canton7** [reply]
FYI @TonyM1958

**Reply by canton7** [reply, marked as answer]
@lamixer Can you give version 1.7.5b1 of the integration a spin, see if it works?

**Reply by lamixer** [reply]
Thanks! Sure.. will attempt this now! You said that the card itself won't need to be updated, only the integration. Does that also apply to foxess_em (which I haven't tried yet but would need to be able to set charge periods) or will that need an update also?

**Reply by canton7** [reply]
The card just reacts to entities being made available by the integration. So yes, the integration needs an udpate but not the card.

foxess_em is separate -- best ask in that repo.

**Reply by TonyM1958** [reply]
@canton7 - so far as I can see, this is the same mapping as H1 and this is just an update to enable these registers for KH in the integration so nothing needs to change in the mapping docs - is that correct?

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/b2e7253d-e3ef-41bb-a32e-8d1837de60b4)

**Reply by canton7** [reply]
Yep, just a note that this has been tested, whereas before we didn't have any documentation or tests saying the KH supported setting charge periods

**Comment by lamixer** [comment]
Wow, great stuff! Thank you.

Ok, I updated FoxESS - Modbus to v 1.7.5b1 and now I have the entities for charge periods and charge/force charge switches and they are showing their correct values!

Naming, for example, is: sensor.foxess_time_period_1_start

If I then install the lovelace card FoxESS - Modbus: Charge Period Card (v1.0.2) and add it to a dashboard, it works!

I did need to restart HA twice for some reason before success.

**Reply by canton7** [reply]
Cool, thanks for checking! I'll promote the beta to a full release.
